ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/Background.hs

```diff
@@ -0,0 +1,344 @@
+    copyThread <- launch $
+      copyToImmDBRunner cdb gcSchedule
+    lgrSnapshotThread <- launch $
+      updateLedgerSnapshotsRunner cdb
+    atomically $ writeTVar cdbBgThreads
+      [gcThread, copyThread, lgrSnapshotThread]
+  where
+    launch = forkLinked cdbThreadRegistry
+
+{-------------------------------------------------------------------------------
+  Copying blocks from the VolatileDB to the ImmutableDB
+-------------------------------------------------------------------------------}
+
+-- | Copy the blocks older than @k@ from the VolatileDB to the ImmutableDB.
+--
+-- These headers of these blocks can be retrieved by dropping the @k@ most
+-- recent blocks from the fragment stored in 'cdbChain'.
+--
+-- The copied blocks are removed from the fragment stored in 'cdbChain'.
+--
+-- This function does not remove blocks from the VolatileDB.
+--
+-- The 'SlotNo' of the tip of the ImmutableDB after copying the blocks is
+-- returned. This can be used for a garbage collection on the VolatileDB.
+--
+-- NOTE: this function would not be safe when called multiple times
+-- concurrently. To enforce thread-safety, a lock is obtained at the start of
+-- this function and released at the end. So in practice, this function can be
+-- called multiple times concurrently, but the calls will be serialised.
+--
+-- NOTE: this function /can/ run concurrently with all other functions, just
+-- not with itself.
+copyToImmDB
+  :: forall m blk.
+     ( MonadSTM   m
+     , MonadCatch m
+     , OuroborosTag (BlockProtocol blk)
+     , HasHeader blk
+     , HasHeader (Header blk)
+     )
+  => ChainDbEnv m blk
+  -> m SlotNo
+copyToImmDB CDB{..} = withCopyLock $ do
+    toCopy <- atomically $ do
+      curChain <- readTVar cdbChain
+      let nbToCopy = max 0 (AF.length curChain - fromIntegral k)
+          toCopy :: [Point blk]
+          toCopy = map headerPoint
+                 $ take nbToCopy
+                 $ AF.toOldestFirst curChain
+      return toCopy
+
+    if null toCopy
+      then trace NoBlocksToCopyToImmDB
+      else forM_ toCopy $ \pt -> do
+        let hash = case pointHash pt of
+              BlockHash h -> h
+              -- There is no actual genesis block that can occur on a chain
+              GenesisHash -> error "genesis block on current chain"
+        -- This call is cheap
+        slotNoAtImmDBTip <- ImmDB.getSlotNoAtTip cdbImmDB
+        if pointSlot pt > slotNoAtImmDBTip
+          then do
+            blk <- VolDB.getKnownBlock cdbVolDB hash
+            -- We're the only one modifying the ImmutableDB, so the tip cannot
+            -- have changed since we last checked it.
+            ImmDB.appendBlock cdbImmDB blk
+            -- TODO the invariant of 'cdbChain' is shortly violated between
+            -- these two lines: the tip was updated on the line above, but the
+            -- anchor point is only updated on the line below.
+            atomically $ removeFromChain pt
```

Comment:
Does this race condition matter?

Downside of doing it one-by-one: all threads watching `cdbChain` are waken up. Not a problem with batching, so that's a performance advantage.

If the race condition doesn't matter after all, there is no reason not to batch (?)